examples/stable_diffusion_v2/ldm/modules/encoders/text_encoder.py

```diff
@@ -172,3 +173,90 @@ def construct(self, text):
         x = x.transpose(1, 0, 2)
         x = self.ln_final(x)
         return x
+
+
+class OpenCLIPResidualAttentionBlock(nn.Cell):
+    def __init__(
+        self,
+        d_model: int,
+        n_head: int,
+        mlp_ratio: float = 4.0,
+        is_cross_attention: bool = False,
+        epsilon: float = 1e-5,
+        use_quick_gelu: bool = False,
+        dtype=ms.float32,
+    ):
+        super().__init__()
+        self.dtype = dtype
+        self.ln_1 = LayerNorm((d_model,), epsilon=epsilon).to_float(self.dtype)
+        self.attn = MultiheadAttention(d_model, n_head, dtype=dtype)
+        if is_cross_attention:
+            self.ln_1_kv = LayerNorm((d_model,), epsilon=epsilon).to_float(self.dtype)
+
+        self.ln_2 = LayerNorm([d_model], epsilon=epsilon).to_float(self.dtype)
+        mlp_width = int(d_model * mlp_ratio)
+        self.mlp = nn.SequentialCell(
+            OrderedDict(
+                [
+                    ("c_fc", nn.Dense(d_model, mlp_width).to_float(self.dtype)),
+                    ("gelu", QuickGELU().to_float(self.dtype) if use_quick_gelu else nn.GELU().to_float(self.dtype)),
+                    ("c_proj", nn.Dense(mlp_width, d_model).to_float(self.dtype)),
+                ]
+            )
+        )
+
+    def construct(self, x, attn_mask):
+        x = x + self.attn(self.ln_1(x), None, None, attn_mask)
+        x = x + self.mlp(self.ln_2(x))
+        return x
+
+
+class OpenCLIPTransformer(nn.Cell):
+    def __init__(self, width, layers, heads, epsilon, use_quick_gelu, dtype=ms.float32):
+        super().__init__()
+        self.width = width
+        self.layers = layers
+        self.resblocks = SequentialCellWithMaskInput(
+            *[
+                OpenCLIPResidualAttentionBlock(
+                    width, heads, epsilon=epsilon, use_quick_gelu=use_quick_gelu, dtype=dtype
+                )
+                for _ in range(layers)
+            ]
+        )
+
+    def construct(self, x, attn_mask):
+        return self.resblocks(x, attn_mask)
+
+
+class OpenClipTextEncoder(nn.Cell):
```

Comment:
how is it different from previous clip text encoder?

Reply:
Basic there is no difference. But the writing style of `OpenClipTextEncoder` strictly follow what stablediffusion does (which relies on `CLIP` and `OpenCLIP` package), thus the migration of the checkpoint is easier in this case